storage/testonly/fake_node_reader.go

```diff
@@ -0,0 +1,196 @@
+// mappings assuming that all the nodes are at a specified tree revision. All the nodeIDs
+// must be distinct.
+func NewFakeNodeReader(mappings []NodeMapping, treeSize, treeRevision int64) *FakeNodeReader {
+	nodeMap := make(map[string]storage.Node)
+
+	for _, mapping := range mappings {
+		_, ok := nodeMap[mapping.NodeID.String()]
+
+		if ok {
+			// Duplicate mapping - bail out
+			return nil
+		}
+
+		nodeMap[mapping.NodeID.String()] = mapping.Node
+	}
+
+	return &FakeNodeReader{nodeMap: nodeMap, treeSize: treeSize, treeRevision: treeRevision}
+}
+
+// GetTreeRevisionAtSize implements the corresponding NodeReader API
+func (f FakeNodeReader) GetTreeRevisionAtSize(treeSize int64) (int64, error) {
+	if f.treeSize != treeSize {
+		return int64(0), fmt.Errorf("GetTreeRevisionAtSize() got treeSize:%d, want: %d", treeSize, f.treeSize)
+	}
+
+	return f.treeRevision, nil
+}
+
+// GetMerkleNodes implements the corresponding NodeReader API
+func (f FakeNodeReader) GetMerkleNodes(treeRevision int64, NodeIDs []storage.NodeID) ([]storage.Node, error) {
+	if f.treeRevision > treeRevision {
+		return []storage.Node{}, fmt.Errorf("GetMerkleNodes() got treeRevision:%d, want up to: %d", treeRevision, f.treeRevision)
+	}
+
+	nodes := make([]storage.Node, 0, len(NodeIDs))
+	for _, nodeID := range NodeIDs {
+		node, ok := f.nodeMap[nodeID.String()]
+
+		if !ok {
+			return []storage.Node{}, fmt.Errorf("GetMerkleNodes() unknown node ID: %v", nodeID)
+		}
+
+		nodes = append(nodes, node)
+	}
+
+	return nodes, nil
+}
+
+func (f FakeNodeReader) hasID(nodeID storage.NodeID) bool {
+	_, ok := f.nodeMap[nodeID.String()]
+	return ok
+}
+
+// MultiFakeNodeReader can provide nodes at multiple revisions. It delegates to a number of
+// FakeNodeReaders, each set up to handle one revision.
+type MultiFakeNodeReader struct {
+	readers []FakeNodeReader
+}
+
+// LeafBatch describes a set of leaves to be loaded into a MultiFakeNodeReader via a compact
+// merkle tree. As each batch is added to the tree a set of node updates are collected
+// and recorded in a FakeNodeReader for that revision.
+type LeafBatch struct {
+	TreeRevision int64
+	Leaves       []string
+	ExpectedRoot string
+}
+
+// NewMultiFakeNodeReader creates a MultiFakeNodeReader delegating to a number of FakeNodeReaders
+func NewMultiFakeNodeReader(readers []FakeNodeReader) *MultiFakeNodeReader {
+	return &MultiFakeNodeReader{readers: readers}
+}
+
+// NewMultiFakeNodeReaderFromLeaves uses a compact Merkle tree to set up the nodes at various
+// revisions. It collates all node updates from a batch of leaf data into one FakeNodeReader.
+// This has the advantage of not needing to manually create all the data structures but the
+// disadvantage is that a bug in the compact tree could be reflected in test using this
+// code. To help guard against this we check the tree root hash after each batch has been
+// processed.
+func NewMultiFakeNodeReaderFromLeaves(batches []LeafBatch) *MultiFakeNodeReader {
```

Comment:
Doc added. Added a check to reject invalid input. Thought about sorting the batches instread but there could still be dupes and the current Go sorting API is onerous.